cpp/include/raft/spectral/lanczos.hpp

```diff
@@ -0,0 +1,1228 @@
+  // -------------------------------------------------------
+
+  while (*iter < maxIter) {
+    ++(*iter);
+
+    // Apply matrix
+    if (shift != 0)
+      CUDA_TRY(cudaMemcpyAsync(
+        lanczosVecs_dev + (*iter) * n, lanczosVecs_dev + (*iter - 1) * n,
+        n * sizeof(ValueType_), cudaMemcpyDeviceToDevice, stream));
+    A->mv(1, lanczosVecs_dev + IDX(0, *iter - 1, n), shift,
+          lanczosVecs_dev + IDX(0, *iter, n));
+
+    // Full reorthogonalization
+    //   "Twice is enough" algorithm per Kahan and Parlett
+    if (reorthogonalize) {
+      CUBLAS_CHECK(cublasgemv(
+        cublas_h, CUBLAS_OP_T, n, *iter, &one, lanczosVecs_dev, n,
+        lanczosVecs_dev + IDX(0, *iter, n), 1, &zero, work_dev, 1, stream));
+
+      CUBLAS_CHECK(cublasgemv(cublas_h, CUBLAS_OP_N, n, *iter, &negOne,
+                              lanczosVecs_dev, n, work_dev, 1, &one,
+                              lanczosVecs_dev + IDX(0, *iter, n), 1, stream));
+
+      CUDA_TRY(cudaMemcpyAsync(alpha_host + (*iter - 1), work_dev + (*iter - 1),
+                               sizeof(ValueType_), cudaMemcpyDeviceToHost,
+                               stream));
+
+      CUBLAS_CHECK(cublasgemv(
+        cublas_h, CUBLAS_OP_T, n, *iter, &one, lanczosVecs_dev, n,
+        lanczosVecs_dev + IDX(0, *iter, n), 1, &zero, work_dev, 1, stream));
+
+      CUBLAS_CHECK(cublasgemv(cublas_h, CUBLAS_OP_N, n, *iter, &negOne,
+                              lanczosVecs_dev, n, work_dev, 1, &one,
+                              lanczosVecs_dev + IDX(0, *iter, n), 1, stream));
+    }
+
+    // Orthogonalization with 3-term recurrence relation
+    else {
+      CUBLAS_CHECK(cublasdot(cublas_h, n,
+                             lanczosVecs_dev + IDX(0, *iter - 1, n), 1,
+                             lanczosVecs_dev + IDX(0, *iter, n), 1,
+                             alpha_host + (*iter - 1), stream));
+
+      auto alpha = -alpha_host[*iter - 1];
+      CUBLAS_CHECK(cublasaxpy(cublas_h, n, &alpha,
+                              lanczosVecs_dev + IDX(0, *iter - 1, n), 1,
+                              lanczosVecs_dev + IDX(0, *iter, n), 1, stream));
+
+      alpha = -beta_host[*iter - 2];
+      CUBLAS_CHECK(cublasaxpy(cublas_h, n, &alpha,
+                              lanczosVecs_dev + IDX(0, *iter - 2, n), 1,
+                              lanczosVecs_dev + IDX(0, *iter, n), 1, stream));
+    }
+
+    // Compute residual
+    CUBLAS_CHECK(cublasnrm2(cublas_h, n, lanczosVecs_dev + IDX(0, *iter, n), 1,
+                            beta_host + *iter - 1, stream));
+
+    // Check if Lanczos has converged
+    if (beta_host[*iter - 1] <= tol) break;
+
+    // Normalize Lanczos vector
+    alpha = 1 / beta_host[*iter - 1];
+    CUBLAS_CHECK(cublasscal(cublas_h, n, &alpha,
+                            lanczosVecs_dev + IDX(0, *iter, n), 1, stream));
+  }
+
+  CUDA_TRY(cudaDeviceSynchronize());
+
+  return 0;
+}
+
+/// Find Householder transform for 3-dimensional system
```

Comment:
I guess this is all done primarily on host because of the tiny data size?

Reply:
Yes.